### ISSUE TITLE
Make Dead Letter Queue current size dependent on current disk usage

### ISSUE THREAD

**Issue by jguay** [elastic/logstash#8794]
- Requested functionality: Ability to delete data from DLQ without restarting logstash.
- Version: 5.6.4 / 6.0.0
- Steps to Reproduce:
1- Set logstash.yml:
```
dead_letter_queue.enable: true
path.dead_letter_queue: "/somepath/dead_letter_queue"
dead_letter_queue.max_bytes: 20mb 
```
2- Fill up the DLQ:
```
PUT testdeadqueue
{
  "mappings": {
    "doc" : {
      "properties": {
        "date": {
          "type": "date"
        }
      }
    }
  }
}
```
logstash conf file : 
```
input {
generator {

}
}
filter {
    ruby {
        code => "event.set('date', rand(36**10).to_s(36))
        "
    }
    mutate {
    remove_field => ["@timestamp", "host", "@version", "sequence", "message"]
    }
    }
output {
elasticsearch {
index => "testdeadqueue"
}
}
```
The following error starts to appear:
```
{"level":"ERROR","loggerName":"org.logstash.common.io.DeadLetterQueueWriter","timeMillis":1511887150437,"thread":"[main]>worker5","logEvent":{"message":"cannot write event to DLQ: reached maxQueueSize of 20971520"}}
```
5- delete the oldest 10Mb file in the dead letter queue folder, still same error continue to come and no new event is recorded in the dead letter queue.
6- Restart logstash will allow 10Mb of DLQ events to be recorded

**Comment by sarfarazahmad89** [comment]
+1

**Comment by schmandforke** [comment]
+1

**Comment by gedge** [comment]
it seems, once logstash thinks that the DLQ has grown too big, it will never write to it again, no matter what you delete, because it never checks the disk

it only ever adds to its internal count of that size: https://github.com/elastic/logstash/blob/master/logstash-core/src/main/java/org/logstash/common/io/DeadLetterQueueWriter.java#L140

hence a restart is required 🤦‍♂

**Comment by chreichert** [comment]
Will this ever be fixed?
Currently the DLQ feature is just not usable as it is.

**Comment by JalehD** [comment]
@jguay should this be closed consider it seems we have the addition of the configuration parameter **dead_letter_queue.flush_interval** in the more recent version of logstash , please see below doc for more information on this.

 https://www.elastic.co/guide/en/logstash/7.13/dead-letter-queues.html#processing-dlq-events

**Comment by vmercierfr** [comment]
According the [source code](https://github.com/elastic/logstash/blob/c698aa224b44c7356fbfeb58ebb0c4a66d72dc71/logstash-core/src/main/java/org/logstash/common/DeadLetterQueueFactory.java#L77), the `dead_letter_queue.flush_interval` parameter control the interval to flush messages from internal queue to disk, but it has no impact on the size of the DLQ queue.

So there is still no parameter to purge DLQ automatically in Logstash 7.13. Unfortunately like @jguay we have to stop our Logstash services to purge the DLQ.

**Comment by roaksoax** [comment]
We believe this issue is now resolved with the 8.4 release, as it introduces a setting to automatically clean-up the DLQ after successfully processing events from it.  See https://www.elastic.co/guide/en/logstash/master/dead-letter-queues.html#auto-clean

Furthermore, there's been further improvements to the DLQ that can be tracked here: https://github.com/elastic/logstash/issues/10493

As such, I';m going to go ahead and close this issue. But, if you believe this issue still applies to you, please re-open it or file a new issue.

Thank you!
